File: go/001.go
func twoSum_1(nums []int, target int) []int {
    size, ret, m := len(nums), []int{}, map[int]int{}
    for i := 0; i < size; i++ {
        if v, ok := m[target-nums[i]]; ok {
            ret = append(ret, v, i)
            return ret
        }
        m[nums[i]] = i
    }
    return ret
}

func twoSum(nums []int, target int) []int {
    size, ret := len(nums), []int{}
    var ns nodes;
    for i, val := range nums {
        ns = append(ns, node{i, val})
    }
    sort.SliceStable(ns, func(i, j int) bool {
        return ns[i].val < ns[j].val
    })
    //sort.Sort(ns)
    for i, j := 0, size - 1; i < j; {
        if ns[i].val + ns[j].val == target {
            if ns[i].index < ns[j].index {
                ret = append(ret, ns[i].index, ns[j].index)
            } else {
                ret = append(ret, ns[j].index, ns[i].index)
            }
            return ret
        }
        if ns[i].val + ns[j].index < target {
            i++
        } else {
            j--
        }
    }
    return ret
}

type node struct {
    index int
    val int
}

type nodes []node

func (n nodes) Less(i, j int) bool {
    return n[i].val < n[j].val
}

func (n nodes) Len() int {
    return len(nodes)
}

func (n nodes) Swap(i, j int) {
    n[i], n[j] = n[j], n[i]
}

File: cpp/001.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        int size = nums.size();
        vector<node> nodes;
        for (int i = 0; i < size; ++i) {
            node temp(i, nums[i]);
            nodes.push_back(temp);
        }
        sort(nodes.begin(), nodes.end(), [](const node &a, const node &b) { return a.val < b.val; });
        vector<int> ret;
        for (int i = 0, j = size - 1; i < j; ) {
            if (nodes[i].val + nodes[j].val == target) {
                int sub1 = nodes[i].index < nodes[j].index ? nodes[i].index : nodes[j].index;
                int sub2 = nodes[i].index < nodes[j].index ? nodes[j].index : nodes[i].index;
                ret.push_back(sub1);
                ret.push_back(sub2);
                return ret;
            }
            if (nodes[i].val + nodes[j].val < target)
                ++i;
            else --j;
        }
        return ret;
    }
    vector<int> twoSum_1(vector<int>& nums, int target) {
        int size = nums.size();
        unordered_map<int, int> m;
        vector<int> ret;
        for (int i = 0; i < size; ++i) {
            if (m.count(target-nums[i])) {
                ret.push_back(m[target-nums[i]]);
                ret.push_back(i);
                return ret;
            }
            m[nums[i]] = i;
        }
        return ret;
    }
private:
    struct node {
        int index;
        int val;
        node(int i, int v): index(i), val(v){}
    };
};
